warehouse/forklift/legacy.py

```diff
@@ -1065,6 +1069,80 @@ def file_upload(request):
+                    "Publishing with GitHub Actions.",
+                )
+            try:
+                attestations = TypeAdapter(list[Attestation]).validate_json(
+                    request.POST["attestations"]
+                )
+            except ValidationError as e:
+                # Log invalid (malformed) attestation upload
+                metrics.increment("warehouse.upload.attestations.malformed")
+                raise _exc_with_message(
+                    HTTPBadRequest,
+                    f"Error while decoding the included attestation: {e}",
+                )
+
+            if len(attestations) > 1:
+                metrics.increment(
+                    "warehouse.upload.attestations." "failed_multiple_attestations"
+                )
+                raise _exc_with_message(
+                    HTTPBadRequest,
+                    "Only a single attestation per-file is supported at the moment.",
+                )
+
+            verification_policy = publisher.publisher_verification_policy(
+                request.oidc_claims
+            )
+            for attestation_model in attestations:
+                try:
+                    # For now, attestations are not stored, just verified
+                    predicate_type, _ = attestation_model.verify(
+                        Verifier.production(),
+                        verification_policy,
+                        Path(temporary_filename),
+                    )
+                except VerificationError as e:
+                    # Log invalid (failed verification) attestation upload
+                    metrics.increment("warehouse.upload.attestations.failed_verify")
+                    raise _exc_with_message(
+                        HTTPBadRequest,
+                        f"Could not verify the uploaded artifact using the included "
+                        f"attestation: {e}",
+                    )
+                except Exception as e:
+                    sentry_sdk.capture_message(
+                        f"Unexpected error while verifying attestation: {e}"
+                    )
+                    raise _exc_with_message(
+                        HTTPBadRequest,
+                        f"Unknown error while trying to verify included "
+                        f"attestations: {e}",
+                    )
+
+                if predicate_type != "https://docs.pypi.org/attestations/publish/v1":
+                    metrics.increment(
+                        "warehouse.upload.attestations."
+                        "failed_unsupported_predicate_type"
+                    )
+                    raise _exc_with_message(
+                        HTTPBadRequest,
+                        f"Attestation with unsupported predicate type: "
+                        f"{predicate_type}",
+                    )
+
+                # Log successful attestation upload
+                metrics.increment("warehouse.upload.attestations.ok")
```

Comment:
Logic LGTM here, but we might want to factor this out into its own helper (since `file_upload` is already quite chunky) 🙂 

Maybe something like `process_attestations`?

Reply:
sounds good, fixed!